Add org.kde.CrowTranslate

Shatur opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [x] Please describe the application briefly.
A simple and lightweight translator that allows you to translate and speak text.
- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project.

The application was previously submitted under [io.crow_translate.CrowTranslate](https://github.com/flathub/io.crow_translate.CrowTranslate), but I need to change the appid due to the transition under the KDE umbrella. After creating a new repository, I will send a PR with a eol rebase manifest to the previous repository. So users will be notified about the appid change.

<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements#control-over-domain
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

bbhtt commented on io.crow_translate.CrowTranslate.yaml:
```diff
@@ -0,0 +1,91 @@
+  - name: crow-translate
+    buildsystem: cmake-ninja
+    sources:
+      - type: git
+        url: https://github.com/crow-translate/crow-translate.git
+        tag: 2.11.1
+        commit: 7113ff01a994584ed0a3ec557fc3a694fcb06b63
+        x-checker-data:
+          type: git
```
This repo is archived, I'm guessing this should be using KDE provided tarballs.

bbhtt replied:
Also is this the source now https://invent.kde.org/office/crow-translate ? (That'd justify the org.kde. ID)

Shatur replied:
> This repo is archived, I'm guessing this should be using KDE provided tarballs.

It's because we don't have a new release yet. But I want to create a flathub entry first, otherwise [the app page](https://apps.kde.org/crowtranslate/) will point to an old release.

> 

> Also is this the source now https://invent.kde.org/office/crow-translate ? (That'd justify the org.kde. ID)

Yes!

travier replied:
Then please use invent as the source, with a fixed commit, not GitHub

Shatur replied:
Switched to using invent. I realized that we imported previous releases, so I can just change the URL and it will work.

bbhtt commented:
cc @flathub/kde

hfiguiere commented:
cc @flathub/kde

travier commented on io.crow_translate.CrowTranslate.yaml:
```diff
@@ -0,0 +1,91 @@
+        cleanup:
+          - /bin
+        sources:
+          - type: git
+            url: https://github.com/tesseract-ocr/tesseract.git
+            tag: 5.3.4
+            commit: 8ee020e14cf5be4e3f0e9beb09b6b050a1871854
```
Please use release tarballs as much as possible, not direct Git repos.

Shatur replied:
Switched to tarballs. I just used previously submitted yaml.

travier commented on io.crow_translate.CrowTranslate.yaml:
```diff
@@ -0,0 +1,91 @@
+  - /lib/cmake
+  - /lib/pkgconfig
+modules:
+  - name: crow-translate
```
This should come last

Shatur replied:
Could you elaborate?

bbhtt replied:
> This should come last

This is the only toplevel module. It's using nested modules.

travier replied:
Ah, good point, I hadn't realized that. Good then

travier commented on io.crow_translate.CrowTranslate.yaml:
```diff
@@ -0,0 +1,91 @@
+            x-checker-data:
+              type: git
+
+      - name: extra-cmake-modules
```
This is already in the SDK

Shatur replied:
Removed!

bbhtt commented:
bot, build org.kde.CrowTranslate

bbhtt commented:
bot, build org.kde.CrowTranslate

bbhtt commented on io.crow_translate.CrowTranslate.yaml:
Manifest filename is wrong `org.kde.CrowTranslate.yaml`

Shatur replied:
Updated, thanks!

bbhtt commented:
Something is wrong with build requests, I'll start it manually once the filename is fixed.

bbhtt commented:
Ah Github is having an outage https://www.githubstatus.com/

Shatur commented:
Sorry, I accidentally calculated SHA1 instead. Fixed.

bbhtt commented:
bot, build org.kde.CrowTranslate

bbhtt commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,73 @@
+    buildsystem: cmake-ninja
+    sources:
+      - type: git
+        url: https://invent.kde.org/office/crow-translate/.git
```
```suggestion
        url: https://invent.kde.org/office/crow-translate.git
```

bbhtt commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,73 @@
+                sha256: ecd7a868403b3963c4e33623595d77f2c87667e2cfdd9b370f87729192061bef
+
+      - name: tessdata_fast
+        disabled: true
```
If you are skipping this why is this added here?

Shatur replied:
I just used what was added here: https://github.com/flathub/io.crow_translate.CrowTranslate/blob/ce7138a8686f7858ad2f99ce03ee11f5b8413f88/io.crow_translate.CrowTranslate.yaml#L70
Didn't even notice that it's disabled. Should I just remove it?

bbhtt replied:
I have no idea if you need it or not. Why was it added to the manifest and why was it disabled?

bbhtt commented:
bot, build org.kde.CrowTranslate

Shatur commented:
Hm... Looks like it's unable to find the .desktop file.
Why it could be?
I generate it [here](https://invent.kde.org/office/crow-translate/-/blob/master/CMakeLists.txt?ref_type=heads#L253).
And install it [here](https://invent.kde.org/office/crow-translate/-/blob/master/CMakeLists.txt?ref_type=heads#L256).

bbhtt commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,61 @@
+  # The app supports capturing screen via KWin and Mutter DBus APIs, but
+  # access to it seems overly broad in spite of the app also supporting XDP API
+  #- --talk-name=org.kde.KWin
+  #- --talk-name=org.gnome.Shell
+rename-icon: crow-translate
```
```suggestion
  #- --talk-name=org.gnome.Shell
rename-desktop-file: io.crow_translate.CrowTranslate.desktop
rename-icon: crow-translate
```

Shatur replied:
But why? Isn't appid is `org.kde.CrowTranslate`?

bbhtt replied:
the desktop file you are installing still has the old name...

Did you build this locally or just checked the 4th checkbox?

Shatur replied:
> the desktop file you are installing still has the old name...

Ah, it's because I using the old commit that corresponds to the version with the old appid. What will be the best way to proceed? Add `rename-desktop-file` for now and remove after the release? Or should I pin to the latest commit?

> Did you build this locally or just checked the 4th checkbox?

I built it on KDE CI, but didn't check after making all these changes, sorry.

bbhtt replied:
> Add rename-desktop-file for now and remove after the release? 

Yea that's fine. We can proceed with the build and test. You can remove it later.

bbhtt commented:
bot, build org.kde.CrowTranslate

bbhtt commented:
I'm guessing the metainfo is still using the old name too, use `rename-appdata-file` similarly in that case.

Shatur commented:
Makes sense, on this commit everything corresponds to the old ID.

bbhtt commented:
bot, build org.kde.CrowTranslate

Shatur commented:
Looks like something changed since the last submission and I need to add more fields. Let me try to run the linter locally.

Shatur commented:
@bbhtt I can update my appdata, but it's located in the repository. Should I try to use the latest commit instead?
Or I can temporarily add it here.

bbhtt commented:
You can add it as a patch https://docs.flatpak.org/en/latest/module-sources.html#patch-sources

bbhtt commented:
bot, build org.kde.CrowTranslate

Shatur commented:
> Tray is missing and icon and tray menu is missing icon on GNOME 46/Fedora 40 VM. I guess this would work in a KDE system.

Yes, I just tested the package on KDE and it works correctly. But it's bad that it doesn't work on GNOME properly. I remember [this](https://github.com/crow-translate/crow-translate/issues/665) issue and installing from PPA helped. Do I need to adjust something in the flatpak configuration to make it work with GNOME?

> Also why is it opening a tray by default? Why not show the window first by default?

It's a good question! I personally keep it enabled because I run the application automatically when the system starts. But the autostart itself is disabled by default. So your suggestion makes total sense. Just disabled "Start minimized" by default.

travier commented:
I don't think it's worth spending time patching the old version. Let's get all the fixes upstream and use the latest commit here even if it's not released yet and we'll move to a release tarball once it's out.

travier commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,64 @@
+runtime: org.kde.Platform
+runtime-version: 5.15-23.08
+finish-args:
+  - --share=ipc
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --device=dri
+  - --socket=pulseaudio
+  - --share=network
+  - --talk-name=org.freedesktop.Notifications
+  - --talk-name=org.kde.StatusNotifierWatcher
```
Let's sort this list

travier commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,64 @@
+  - --share=network
+  - --talk-name=org.freedesktop.Notifications
+  - --talk-name=org.kde.StatusNotifierWatcher
+  # The app supports capturing screen via KWin and Mutter DBus APIs, but
+  # access to it seems overly broad in spite of the app also supporting XDP API
+  #- --talk-name=org.kde.KWin
+  #- --talk-name=org.gnome.Shell
```
This should go into a README for the repo to make it easier to find.

Maybe also in a message in the application when running as Flatpak.

Shatur replied:
Removed completely. `org.kde.KWin` and `org.gnome.Shell` aren't necessary.

travier commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,64 @@
+          - -DBUILD_SHARED_LIBS:BOOL=ON
+        cleanup:
+          - /bin
+        sources:
```
You should add `x-checker-data` config for all sources

Shatur replied:
I checked other projects and looks like they don't use `x-checker-data` for archives from GitHub.
For example, [org.flathub.flatpak-external-data-checker](https://github.com/flathub/org.flathub.flatpak-external-data-checker/blob/502b06d4d4e67372ee1c699be27d920995c70e90/org.flathub.flatpak-external-data-checker.yml#L159C27-L159C34) does not do this.

bbhtt commented:
> Do I need to adjust something in the flatpak configuration to make it work with GNOME?

Not it should work with the current configuration? What icon name are using for the tray? The icon needs to be exported by Flatpak on host so that it is available in XDG_DATA_DIRS. The tray extension cannot access an icon that Flatpak hasn't exported.

Flatpak will export icons from `$prefix/share/icons/hicolor/...` with the following patterns `$FLATPAK_ID, $FLATPAK_ID.foo, $FLATPAK_ID-foo` .png/.svg

Shatur commented:
Ah, I use `crow-translate` as the icon. That's the issue.
Will change it switch to the latest master.

Shatur commented:
Done! Thank you a lot for the feedback. I'm new to Flatpak and looks like the app wasn't properly packaged before.
But it should be good now. I switched to the master and fixed the mentioned issues.

bbhtt commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,87 @@
+  - --talk-name=org.kde.StatusNotifierWatcher
+copy-icon: true
+command: crow
+separate-locales: false
```
Why is this disabled?

Shatur replied:
The is unable to load localization without this line.

bbhtt replied:
Are you sure? You need to install the `.Locale` extension for that locale. By default Flatpak will only install the locale extension for your host locale.

This disabled creating locale extension and bundles all locales into the app.

Shatur replied:
Ah, I didn't know about it! In my application I have locale selection and it could be confusing when user picks a localization and it doesn't work.
Should I keep `separate-locales: false` or remove it?

bbhtt replied:
Then keep it, usually apps follow the system locale.

bbhtt commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,87 @@
+runtime: org.kde.Platform
+runtime-version: 5.15-23.08
+add-extensions:
+  org.kde.CrowTranslate.tessdata:
```
What is this for?

Shatur replied:
It's needed for tesseract models for screen reading. I used [this](https://github.com/flathub/io.github.manisandro.gImageReader/blob/master/io.github.manisandro.gImageReader.yml) app, which also uses Tesseract, as a reference.

Shatur commented:
Anything else needed from me?
Should we wait for the release drop before creating a repository?

travier commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,87 @@
+                  project-id: 1549
+                  url-template: https://github.com/DanBloomberg/leptonica/releases/download/$version/leptonica-$version.tar.gz
+
+      # The `tessdata_fast` variant is recommended for end users: https://tesseract-ocr.github.io/tessdoc/Data-Files
```
Can you move that to a README file?

travier replied:
Ah, this copied from https://github.com/flathub/io.github.manisandro.gImageReader/blob/master/io.github.manisandro.gImageReader.yml?rgh-link-date=2024-07-09T12%3A52%3A57Z but I don't know how this works.

Shatur replied:
Done!

travier commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,87 @@
+          - cp -av * ${FLATPAK_DEST}/share/tessdata/
+        sources:
+          - type: git
+            url: https://github.com/tesseract-ocr/tessdata_fast
```
Can this be fetched from a tarball instead of Git?

Shatur replied:
Fixed, thanks!

hfiguiere commented on org.kde.CrowTranslate.yaml:
```diff
@@ -0,0 +1,87 @@
+add-extensions:
+  org.kde.CrowTranslate.tessdata:
+    directory: share/tessdata
+    version: stable
+    versions: stable;beta;test
```
but then there is no such package. not even in the submission queue.

Shatur replied:
Yes, I copied it from [this](https://github.com/flathub/io.github.manisandro.gImageReader/blob/master/io.github.manisandro.gImageReader.yml) app.

> but then there is no such package. not even in the submission queue.

Ah, I thought that it creates a extension for the app. Do I need to package tessdata separately? I don't see ` io.github.manisandro.gImageReader.tessdata` anywhere as well...

I'm new to flatpak.

hfiguiere replied:
two wrongs don't make a right.

travier replied:
Let's try to figure out how to do properly. I'm not familiar with this so we'll have to search in the docs.

hfiguiere replied:
`man flatpak-metadata`

[...]

```
       version (string)
           The branch to use when looking for the extension. If this is not specified, it defaults to the branch of the application
           or runtime that the extension point is for. Available since 0.4.1.

       versions (string)
           The branches to use when looking for the extension. If this is not specified, it defaults to the branch of the
           application or runtime that the extension point is for. Available since 0.9.1, and backported to the 0.8.x branch in
           0.8.4.

```

hfiguiere replied:
What it doesn't say is that they are equivalent one being the plural form.

That doesn't make an extension being built. That just declare the extension point.

`man flatpak-manifest`

```
       bundle (boolean)
           If this is true, then the data created in the extension directory is omitted from the result, and instead packaged in a
           separate extension.

```

Shatur replied:
> Let's try to figure out how to do properly

Reading it right now. Looks like I need to create a separate manifest for it and add a special entry to my app metadata.
We probably need to notify the maintainer of `io.github.manisandro.gImageReader` as well about it.

> man flatpak-metadata

We aren't talking about the version string, but about the whole `add-extensions` entry.

hfiguiere replied:
> Reading it right now. Looks like I need to create a separate manifest for it and add a special entry to my app metadata.

no.

Shatur replied:
> That doesn't make an extension being built. That just declare the extension point.

It's because `bundle` is false by default?
So instead of setting a version and creating a separate manifest, I can replace `version` with `bundle: true`?

bbhtt replied:
> Yes, I copied it from [this](https://github.com/flathub/io.github.manisandro.gImageReader/blob/master/io.github.manisandro.gImageReader.yml) app.

See the tip here https://docs.flatpak.org/en/latest/extension.html#extension-point Since you don't have a beta application you don't need this. `test` is not needed, it's a temporary repo.

> It's because bundle is false by default?

Yes

https://docs.flatpak.org/en/latest/extension.html#bundled-extensions

Shatur replied:
I think I figured it out. I enabled bundling and now the extension creates properly. I also set version that matches the tessdata version.

Opened an issue no notify the maintainer of io.github.manisandro.gImageReader about it: https://github.com/flathub/io.github.manisandro.gImageReader/issues/70

@hfiguiere Thanks for pointing it out, but if your goal is to help the community, try to give helpful advice to newcomers instead of just saying "it's wrong".

Shatur replied:
That's what I thought, thanks!
It looks like, in this case, the tessada was installed as part of the main manifest, but the extension relied on a separate, non-existent manifest. The author has just applied the same fix: https://github.com/flathub/io.github.manisandro.gImageReader/pull/71

travier commented:
> Anything else needed from me? Should we wait for the release drop before creating a repository?

We can "merge" this once this is ready, which is once the @flathub/kde team is giving it 👍🏻 and then we'll ask Flathub Admins to trigger the repo creation.

bbhtt commented:
bot, build org.kde.CrowTranslate

bbhtt commented:
The tray menu items are still missing icons

![image](https://github.com/flathub/flathub/assets/62639087/cd8d38c4-9be4-4be6-add2-c8fdf623056a)

Shatur commented:
Thanks for testing!

I wondering why it happens... I use `window`, `configure` and `application-exit` icons for them respectively. It works in KDE. Are they missing in Advaita?

bbhtt commented:
The first two doesn't exist.

bbhtt commented:
You can probably use preferences-other for the configure, it seems to exist in Breeze and Adwaita legacy. The window one doesn't have an option that exists in both other than window-new.

Shatur commented:
Thanks! Used `preferences-other` for `configure` and `window-maximize` for `window`.

Shatur commented:
Switched to just using the application tray icon instead of `window-maximize`, I think it fits better.

bbhtt commented:
bot, build org.kde.CrowTranslate

travier commented:
bot, build org.kde.CrowTranslate

travier commented:
Works well for me. LGTM

Shatur commented:
> After creating a new repository, I will send a PR with a eol rebase manifest to the previous repository. So users will be notified about the appid change.

Should I proceed with this?

bbhtt commented:
Yea
